src/lib.rs

```diff
@@ -1,39 +1,76 @@
-//! A fast, small, full-featured async-aware oneshot channel
-//!
-//! Unique feature: wait for receiver to be waiting.
-//!
-//! Also supports the full range of things you'd expect.
+//! An easy-to-use, high-performance single-message async channel for
+//! a single sender/receiver pair.
 #![no_std]
+
+#[cfg(feature="alloc")]
 extern crate alloc;
-use alloc::sync::Arc;
 
-mod inner;
-pub(crate) use inner::Inner;
+#[cfg(feature="alloc")]
+use alloc::boxed::Box;
+use core::{pin::Pin, ptr::NonNull};
+
+// Things shared between the sender and receiver
+mod shared;
+pub use shared::{Closed, Hatch};
+use shared::*;
+
+pub mod sender;
+pub mod receiver;
+
+pub use sender::SendError;
 
-mod sender;
-pub use sender::Sender;
+/// This is a convenience alias where the lifetime is static.
+pub type Receiver<T> = receiver::Receiver<'static, T>;
+/// This is a convenience alias where the lifetime is static.
+pub type Sender<T> = sender::Sender<'static, T>;
 
-mod receiver;
-pub use receiver::Receiver;
+#[cfg(feature="alloc")]
+/// Creates a new hatch backed by a box. Returns a [`Sender`]/[`Receiver`] pair.
+pub fn hatch<T>() -> (Sender<T>, Receiver<T>) {
+    let unbox = Box::leak(Box::new(Hatch::default()));
+    let non = unsafe { NonNull::new_unchecked(unbox) };
+    let holder = Holder::SharedBoxPtr(non);
+    // Safe because we have exclusive access
+    unsafe { (Sender::new(holder), Receiver::new(holder)) }
+}
+
+/// Creates a new hatch backed by a ref to an existing Hatch. Unlike
+/// [`hatch`], this is not allocated on the stack and is bound by the
+/// lifetime of the passed mut ref.
+pub fn ref_hatch<T>(
+    hatch: Pin<&mut Hatch<T>>
+) -> (sender::Sender<T>, receiver::Receiver<T>) {
```

Comment:
```suggestion
) -> (sender::Sender<'_, T>, receiver::Receiver<'_, T>) {
```
implicit lifetimes are deprecated.